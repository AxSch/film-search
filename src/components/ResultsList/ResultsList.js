import React from 'react'
import { useDispatch } from 'react-redux'
import { setCurrentItem } from '../../reducers/searchResults/searchResultSlice'
import Result from '../Result/Result'
import { Dropdown, Suggestions, ResultsSection, ResultsHeading } from './ResultsList.styled'

const ResultsList = ({ results, handleDropdown, filter, isDropdown }) => {
    const dispatch = useDispatch()

    const handleOnClick = (result) => {
        handleDropdown(false)
        dispatch(setCurrentItem(result))
    }

    const renderTVshows = (results, filter) => {
        if (filter === 'all' || filter === 'tvShows') {
            return (
                <ResultsSection data-testid="tvshowsSection">
                    <ResultsHeading>
                        <span>TV Shows</span>
                    </ResultsHeading>
                    <div>
                        {results['tvShows'].map((tvShow, index) => {
                            return (
                                <div onClick={() => handleOnClick(tvShow)} key={index}>
                                    <Result category="tvShows" resultData={tvShow} />
                                </div>
                            )
                        })}
                    </div>
                </ResultsSection>
            )
        }
    }

    const renderMovies = (results, filter) => {
        if (filter === 'all' || filter === 'movies') {
            return (
                <ResultsSection data-testid="moviesSection">
                    <ResultsHeading>
                        <span>Movies</span>
                    </ResultsHeading>
                    <div>
                        {results['movies'].map((movie, index) => {
                            return (
                                <div onClick={() => handleOnClick(movie)} key={index}>
                                    <Result category="movies" resultData={movie} />
                                </div>
                            )
                        })}
                    </div>
                </ResultsSection>
            )
        }
    }

    const renderPeople = (results, filter) => {
        if (filter === 'all' || filter === 'people') {
            return (
                <ResultsSection isPeople={true} data-testid="peopleSection">
                    <ResultsHeading>
                        <span>People</span>
                    </ResultsHeading>
                    <div>
                        {results['people'].map((people, index) => {
                            return (
                                <div onClick={() => handleOnClick(people)} key={index}>
                                    <Result category="people" resultData={people} isDropdown={isDropdown} />
                                </div>
                            )
                        })}
                    </div>
                </ResultsSection>
            )
        }
    }

    const renderResults = () => {
        return (
            <>
                {isDropdown ? <Suggestions data-testid="suggestions">Suggestions</Suggestions> : <h3 data-testid="results">Results</h3>}
                {renderTVshows(results, filter)}
                {renderMovies(results, filter)}
                {renderPeople(results, filter)}
            </>
        )
    }
    return (
        <Dropdown isDropdown={isDropdown} data-testid="resultsList">
            {renderResults()}
        </Dropdown>
    )
}

export default ResultsList
