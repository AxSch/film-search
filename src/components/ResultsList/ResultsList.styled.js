import styled from 'styled-components'

const Dropdown = styled.ul`
    margin: 1.2rem 0;
    list-style-type: none;
    padding-left: 0;
    border-radius: 4px;
    padding: ${props => (props.isDropdown ? `1.2rem` : 0)};
    box-shadow: ${props => (props.isDropdown ? `0px 0px 0px 1px rgba(0, 0, 0, 0.2)` : '')};
`

const Suggestions = styled.div`
    margin-bottom: 1.2rem;
    font-weight: 500;
`

const ResultsSection = styled.div`
    display: flex;
    margin: 1.1rem 0;
    padding: 1.1rem 0;
    border-bottom: ${props => (!props.isPeople ? `1px solid rgba(0, 0, 0, 0.2)` : '')};
`

const ResultsHeading = styled.div`
    width: 30%;
`

export {
    Dropdown,
    Suggestions,
    ResultsSection,
    ResultsHeading
}
