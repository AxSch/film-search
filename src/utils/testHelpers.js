import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom'
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import searchResultReducer from '../reducers/searchResults/searchResultSlice';

export const renderWithRouter = (ui, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route)
  
    return render(ui, { wrapper: BrowserRouter })
  }


export const renderConnected = (
  ui, {
    store = createStore(searchResultReducer, {}),
    route = '/',
    ...renderOptions
  } = {}
) => {
  window.history.pushState({}, 'Test page', route)
  const Wrapper = ({ children }) => (
    <Provider store={store}>
      <BrowserRouter>
        {children}
      </BrowserRouter>
    </Provider>
  );
  return render(ui, { wrapper: Wrapper, ...renderOptions});
};

export default renderConnected;