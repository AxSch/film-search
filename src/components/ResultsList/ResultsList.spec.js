import React from 'react'
import { renderConnected } from '../../utils/testHelpers';
import ResultsList from './ResultsList'

const initialState = {
    results: {},
    currentItem: {}
}

const mockResults = {
    movies: [
        {
            adult: false,
            backdrop_path: null,
            genre_ids: [10402],
            id: 413131,
            original_language: "en",
            original_title: "Will Smith: Live in Concert",
            overview: "Will Smith performs live in concert some of his old and new hits.",
            popularity: 2.173,
            poster_path: "/3jfOK5O8HEa6F8cRLVmLTm0YlE8.jpg",
            release_date: "",
            title: "Will Smith: Live in Concert",
            video: false,
            vote_average: 6.4,
            vote_count: 7,
        },
        {
            adult: false,
            backdrop_path: null,
            genre_ids: [10402],
            id: 413131,
            original_language: "en",
            original_title: "Will Smith - Goes to New York",
            overview: "Will Smith goes to the big apple.",
            popularity: 0.9,
            poster_path: "/3jfOK5O8HEa6F8cRLVmLTm0YlE8.jpg",
            release_date: "",
            title: "Will Smith - Goes to New York",
            video: false,
            vote_average: 2.2,
            vote_count: 10,
        },
    ],
    tvShows: [
        {
            backdrop_path: "/mygD13dQe2sTQ7EMEx065DB1qmD.jpg",
            first_air_date: "2019-02-27",
            genre_ids: (2) [10764, 99],
            id: 87732,
            name: "Will Smith's Bucket List",
            origin_country: ["US"],
            original_language: "en",
            original_name: "Will Smith's Bucket List",
            overview: "There arenâ€™t many people in the world who have a bucket list quite like Will Smith. Now you can join him on his unbelievable adventure as he travels the globe, takes on insane challenges, overcomes obstacles and punches fear in the mouth!",
            popularity: 4.79,
            poster_path: "/8FAd59PHpLzIjMQDk3AIrmEQB0K.jpg",
            vote_average: 6,
            vote_count: 4
        }
    ],
    people: [
        {
            adult: false,
            gender: 2,
            id: 2888,
            known_for: [{}, {}, {}],
            known_for_department: "Acting",
            name: "Will Smith",
            popularity: 17.466,
            profile_path: "/eze9FO9VuryXLP0aF2cRqPCcibN.jpg"
        }
    ]
}


describe('ResultsList', () => {
    let container, getByTestId, queryByTestId
    
    beforeEach(() => {
        
    })

    it('should render successfully', () => {
        const utils = renderConnected(<ResultsList results={mockResults} handleDropdown={() => {}}  filter='all' isDropdown={false} />, { initialState });
        container = utils.container;
        getByTestId = utils.getByTestId;
        expect(getByTestId("resultsList")).toBeInTheDocument();
        expect(getByTestId("results")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should render all results - filter set to all', () => {
        const utils = renderConnected(<ResultsList results={mockResults} handleDropdown={() => {}}  filter='all' isDropdown={false} />, { initialState });
        container = utils.container;
        getByTestId = utils.getByTestId;

        expect(getByTestId("resultsList")).toBeInTheDocument();
        expect(getByTestId("peopleSection")).toBeInTheDocument();
        expect(getByTestId("tvshowsSection")).toBeInTheDocument();
        expect(getByTestId("moviesSection")).toBeInTheDocument();
    })

    it('should render all results - filter set to people', () => {
        const utils = renderConnected(<ResultsList results={mockResults} handleDropdown={() => {}}  filter='people' isDropdown={false} />, { initialState });
        queryByTestId = utils.queryByTestId;

        expect(queryByTestId("peopleSection")).toBeInTheDocument();
        expect(queryByTestId("tvshowsSection")).not.toBeInTheDocument();
        expect(queryByTestId("moviesSection")).not.toBeInTheDocument();
    })

    it('should render all results - filter set to movies', () => {
        const utils = renderConnected(<ResultsList results={mockResults} handleDropdown={() => {}}  filter='movies' isDropdown={false} />, { initialState });
        queryByTestId = utils.queryByTestId;

        expect(queryByTestId("moviesSection")).toBeInTheDocument();
        expect(queryByTestId("peopleSection")).not.toBeInTheDocument();
        expect(queryByTestId("tvshowsSection")).not.toBeInTheDocument();
    })

    it('should render all results - filter set to tvShows', () => {
        const utils = renderConnected(<ResultsList results={mockResults} handleDropdown={() => {}}  filter='tvShows' isDropdown={false} />, { initialState });
        queryByTestId = utils.queryByTestId;

        expect(queryByTestId("tvshowsSection")).toBeInTheDocument();
        expect(queryByTestId("moviesSection")).not.toBeInTheDocument();
        expect(queryByTestId("peopleSection")).not.toBeInTheDocument();
    })

    it('should render suggestions', () => {
        const utils = renderConnected(<ResultsList results={mockResults} handleDropdown={() => {}}  filter='all' isDropdown={true} />, { initialState });
        getByTestId = utils.getByTestId;

        expect(getByTestId("suggestions")).toBeInTheDocument();
        expect(queryByTestId("tvshowsSection")).toBeInTheDocument();
        expect(queryByTestId("moviesSection")).toBeInTheDocument();
        expect(queryByTestId("peopleSection")).toBeInTheDocument();
    })

})