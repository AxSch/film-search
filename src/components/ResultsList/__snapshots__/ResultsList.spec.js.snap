// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsList should render successfully 1`] = `
<div>
  <ul
    class="sc-dlnjwi iajivG"
    data-testid="resultsList"
  >
    <h3
      data-testid="results"
    >
      Results
    </h3>
    <div
      class="sc-eCApnc hxHobU"
      data-testid="tvshowsSection"
    >
      <div
        class="sc-jSFjdj cqLFpO"
      >
        <span>
          TV Shows
        </span>
      </div>
      <div>
        <div>
          <li
            class="sc-bdnxRM crZWod"
            data-testid="resultItem"
          >
            <a
              href="/item/87732&category=tvShows"
            >
              <span
                class="sc-gtsrHT kiexgk"
                data-testid="resultName"
              >
                Will Smith's Bucket List
              </span>
            </a>
          </li>
        </div>
      </div>
    </div>
    <div
      class="sc-eCApnc hxHobU"
      data-testid="moviesSection"
    >
      <div
        class="sc-jSFjdj cqLFpO"
      >
        <span>
          Movies
        </span>
      </div>
      <div>
        <div>
          <li
            class="sc-bdnxRM crZWod"
            data-testid="resultItemMovies"
          >
            <a
              href="/item/413131&category=movies"
            >
              <span>
                Will Smith: Live in Concert
              </span>
            </a>
          </li>
        </div>
        <div>
          <li
            class="sc-bdnxRM crZWod"
            data-testid="resultItemMovies"
          >
            <a
              href="/item/413131&category=movies"
            >
              <span>
                Will Smith - Goes to New York
              </span>
            </a>
          </li>
        </div>
      </div>
    </div>
    <div
      class="sc-eCApnc ffJBIP"
      data-testid="peopleSection"
    >
      <div
        class="sc-jSFjdj cqLFpO"
      >
        <span>
          People
        </span>
      </div>
      <div>
        <div>
          <li
            class="sc-bdnxRM crZWod"
            data-testid="resultItem"
          >
            <a
              href="/item/2888&category=people"
            >
              <span
                class="sc-gtsrHT kiexgk"
                data-testid="resultName"
              >
                Will Smith
              </span>
              <span>
                - 
                Acting
              </span>
            </a>
          </li>
        </div>
      </div>
    </div>
  </ul>
</div>
`;
